File: backend/app/schemas/quiz.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.quiz_attempt import AttemptStatus


# ── Request ───────────────────────────────────────────────────────────────────

class QuizGenerateRequest(BaseModel):
    document_id: int
    num_questions: int = Field(ge=5, le=50, description="Number of questions to generate (5–50)")
    difficulty: str = Field(pattern="^(easy|medium|hard)$", description="easy | medium | hard")


class AnswerCreate(BaseModel):
    text: str = Field(min_length=1)
    is_correct: bool
    order: int


class QuestionCreate(BaseModel):
    type: str = Field(pattern="^(MCQ|TrueFalse|ShortAnswer)$")
    text: str = Field(min_length=1)
    order: int
    points: int = Field(default=1, ge=1, description="Point value for this question")
    answers: list[AnswerCreate] = []


class QuizCreateRequest(BaseModel):
    class_id: Optional[uuid.UUID] = None
    title: str = Field(min_length=1, max_length=255)
    description: Optional[str] = None
    difficulty: Optional[str] = Field(None, pattern="^(easy|medium|hard)$")
    duration_minutes: Optional[int] = Field(None, ge=1)
    max_attempts: Optional[int] = Field(None, ge=1)
    questions: list[QuestionCreate] = []


class QuizUpdateRequest(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    status: Optional[str] = Field(None, pattern="^(draft|published|archived)$")
    difficulty: Optional[str] = Field(None, pattern="^(easy|medium|hard)$")
    duration_minutes: Optional[int] = Field(None, ge=1)
    max_attempts: Optional[int] = Field(None, ge=1)


# ── Job responses ─────────────────────────────────────────────────────────────

class QuizJobResponse(BaseModel):
    """Returned immediately after POST /quiz/generate."""
    job_id: uuid.UUID
    status: str
    document_id: int
    num_questions: int
    difficulty: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class QuizJobDetailResponse(BaseModel):
    """Returned by GET /quiz/jobs/{job_id}."""
    job_id: uuid.UUID
    status: str
    document_id: int
    quiz_id: Optional[int]
    num_questions: int
    difficulty: str
    error_message: Optional[str]
    total_tokens: Optional[int]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Quiz detail responses ─────────────────────────────────────────────────────

class AnswerResponse(BaseModel):
    id: int
    text: str
    is_correct: bool
    order: int

    model_config = ConfigDict(from_attributes=True)


class QuestionResponse(BaseModel):
    id: int
    type: str
    text: str
    order: int
    points: int
    answers: list[AnswerResponse]

    model_config = ConfigDict(from_attributes=True)


class QuizResponse(BaseModel):
    id: int
    title: str
    description: Optional[str]
    difficulty: Optional[str]
    document_id: Optional[int]
    class_id: Optional[uuid.UUID]
    status: str
    duration_minutes: Optional[int]
    max_attempts: Optional[int]
    questions: list[QuestionResponse]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Quiz attempt schemas ───────────────────────────────────────────────────────

class QuizAttemptStartResponse(BaseModel):
    attempt_id: int
    quiz_id: int
    status: AttemptStatus
    started_at: datetime
    # None when the quiz has no time limit
    expires_at: Optional[datetime]

    model_config = ConfigDict(from_attributes=True)


class StudentAnswerSubmit(BaseModel):
    question_id: int
    # Free text for ShortAnswer; answer text for MCQ/TrueFalse
    answer_text: Optional[str] = None


class QuizSubmitRequest(BaseModel):
    answers: list[StudentAnswerSubmit] = []


class QuizAttemptSubmitResponse(BaseModel):
    attempt_id: int
    quiz_id: int
    status: AttemptStatus
    submitted_at: datetime
    total_questions: int
    answers_recorded: int

    model_config = ConfigDict(from_attributes=True)

File: frontend/types/quiz.ts
export type QuizDifficulty = 'easy' | 'medium' | 'hard';
export type QuizStatus = 'draft' | 'published' | 'archived';
export type QuestionType = 'MCQ' | 'TrueFalse' | 'ShortAnswer';
export type JobStatus = 'pending' | 'processing' | 'completed' | 'failed';

// ── API response shapes (mirror backend schemas) ──────────────────────────────

export interface AnswerResponse {
  id: number;
  text: string;
  is_correct: boolean;
  order: number;
}

export interface QuestionResponse {
  id: number;
  type: QuestionType;
  text: string;
  order: number;
  answers: AnswerResponse[];
}

export interface QuizResponse {
  id: number;
  title: string;
  description: string | null;
  difficulty: QuizDifficulty | null;
  document_id: number | null;
  class_id: string | null;
  status: QuizStatus;
  duration_minutes: number | null;
  max_attempts: number | null;
  questions: QuestionResponse[];
  created_at: string;
  updated_at: string;
}

export interface QuizJobResponse {
  job_id: string;
  status: JobStatus;
  document_id: number;
  num_questions: number;
  difficulty: QuizDifficulty;
  created_at: string;
}

export interface QuizJobDetailResponse {
  job_id: string;
  status: JobStatus;
  document_id: number;
  quiz_id: number | null;
  num_questions: number;
  difficulty: QuizDifficulty;
  error_message: string | null;
  total_tokens: number | null;
  created_at: string;
  updated_at: string;
}

// ── Client-side editing shapes ────────────────────────────────────────────────

export interface LocalAnswer {
  localId: string; // temp client key
  id?: number;
  text: string;
  is_correct: boolean;
  order: number;
}

export interface LocalQuestion {
  localId: string; // temp client key
  id?: number;
  type: QuestionType;
  text: string;
  order: number;
  answers: LocalAnswer[];
}

export interface QuizFormData {
  title: string;
  description: string;
  duration_minutes: string;
  max_attempts: string;
  difficulty: QuizDifficulty;
}

// ── Request shapes ────────────────────────────────────────────────────────────

export interface QuizGenerateRequest {
  document_id: number;
  num_questions: number;
  difficulty: QuizDifficulty;
}

export interface QuizCreateRequest {
  class_id?: string;
  title: string;
  description?: string;
  difficulty?: QuizDifficulty;
  duration_minutes?: number;
  max_attempts?: number;
  questions: Array<{
    type: QuestionType;
    text: string;
    order: number;
    answers: Array<{ text: string; is_correct: boolean; order: number }>;
  }>;
}

export interface QuizUpdateRequest {
  title?: string;
  description?: string;
  status?: QuizStatus;
  difficulty?: QuizDifficulty;
  duration_minutes?: number;
  max_attempts?: number;
}

// ── Attempt shapes ────────────────────────────────────────────────────────────

export type AttemptStatus = 'started' | 'in_progress' | 'submitted';

export interface QuizAttemptStartResponse {
  attempt_id: number;
  quiz_id: number;
  status: AttemptStatus;
  started_at: string;
  expires_at: string | null;
}

export interface StudentAnswerSubmit {
  question_id: number;
  answer_text: string | null;
}

export interface QuizSubmitRequest {
  answers: StudentAnswerSubmit[];
}

export interface QuizAttemptSubmitResponse {
  attempt_id: number;
  quiz_id: number;
  status: AttemptStatus;
  submitted_at: string;
  total_questions: number;
  answers_recorded: number;
}

// ── Results & grading ─────────────────────────────────────────────────────────

export interface QuestionResult {
  question_id: number;
  question_text: string;
  question_type: QuestionType;
  points: number;
  student_answer: string | null;
  correct_answer: string | null;
  is_correct: boolean | null;
  score: number | null;
  needs_review: boolean;
  feedback_text: string | null;
  key_points: string[];
  improvement_suggestion: string | null;
  llm_reasoning: string | null;
  bleu_score: number | null;
  rouge_l_score: number | null;
  teacher_score: number | null;
}

export interface AttemptResultResponse {
  attempt_id: number;
  quiz_id: number;
  quiz_title: string;
  quiz_difficulty: string | null;
  status: string;
  started_at: string;
  submitted_at: string | null;
  duration_seconds: number | null;
  score: number | null;
  class_average: number | null;
  total_questions: number;
  auto_graded: number;
  pending_review: number;
  feedback_generated: boolean;
  questions: QuestionResult[];
  total_tokens_used: number;
}

// ── Quiz history ──────────────────────────────────────────────────────────────

export interface QuizHistoryItem {
  attempt_id: number;
  quiz_id: number;
  quiz_title: string;
  quiz_difficulty: string | null;
  class_id: string | null;
  class_name: string | null;
  score: number | null;
  status: string;
  started_at: string;
  submitted_at: string | null;
}

export interface PaginatedQuizHistory {
  items: QuizHistoryItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// ── Student assigned quizzes ──────────────────────────────────────────────────

export interface AssignedByInfo {
  id: string;
  email: string;
}

export interface StudentAssignedQuizItem {
  assignment_id: string;
  assignment_status: string;
  assigned_at: string;
  assigned_by: AssignedByInfo;
  due_date: string | null;
  class_id: string;
  class_name: string;
  quiz: QuizResponse;
  attempt_id: number | null;
  attempt_status: 'started' | 'in_progress' | 'submitted' | null;
  score: number | null;
}

// ── Teacher quiz list ─────────────────────────────────────────────────────────

export interface TeacherQuizListItem {
  id: number;
  title: string;
  description: string | null;
  status: QuizStatus;
  difficulty: QuizDifficulty | null;
  class_id: string | null;
  class_name: string | null;
  question_count: number;
  attempt_count: number;
  avg_score: number | null;
  created_at: string;
}

export interface TeacherQuizListResponse {
  items: TeacherQuizListItem[];
  total: number;
}

// ── Quiz attempt list (teacher view) ─────────────────────────────────────────

export interface QuizAttemptItem {
  attempt_id: number;
  student_id: string;
  student_email: string;
  student_name: string | null;
  status: AttemptStatus | 'submitted';
  score: number | null;
  started_at: string;
  submitted_at: string | null;
}

export interface QuizAttemptsListResponse {
  quiz_id: number;
  quiz_title: string;
  items: QuizAttemptItem[];
  total: number;
}
